tests/flow/test_foreach.py

```diff
@@ -0,0 +1,491 @@
+        self.env.assertEquals(res.nodes_deleted, 0)
+
+        # validate that the end state is correct
+        # there should be one node in the graph, with label N, 
+        # properties: {v: 2, name: 'RAZmon', li: [1] and 4 nodes with label M
+        # and properties: {v: 1, name: 'raz'} and li from 1 to 4
+        res = graph.query("MATCH (n:N) return n")
+        n = Node(label='N', properties={'v': 1, 'name': 'RAZ', 'li': [1, 2, 3, 4]})
+        self.env.assertEquals(len(res.result_set), 1)
+        self.env.assertEquals(res.result_set[0][0], n)
+        res = graph.query("MATCH (m:M) return m")
+        self.env.assertEquals(len(res.result_set), 4)
+        for i in range(len(res.result_set)):
+            m = Node(label= 'M', properties={'v': 1, 'name': 'raz', 'li': [i+1]})
+            self.env.assertEquals(res.result_set[i][0], m)
+
+        # clear nodes with label M
+        res = graph.query("MATCH (m:M) DELETE m")
+        self.env.assertEquals(res.nodes_deleted, 4)
+
+        # ----------------------------------------------------------------------
+        # triple embedding of a FOREACH clause, referencing node properties
+        # ----------------------------------------------------------------------
+
+        query = """
+        MATCH (n:N)
+        WITH collect(n) as ns
+        FOREACH(n in ns |
+            FOREACH(x in n.li |
+                FOREACH(do_action IN CASE WHEN x > 2 THEN [1] ELSE [] END |
+                    CREATE (t:TEMP {v: 2 * x + n.v, name: toLower(n.name), li: n.li + [x]})
+                )
+            )
+        )
+        """
+
+        res = graph.query(query)
+
+        # validate the query statistics
+        self.env.assertEquals(res.nodes_created, 2)
+        self.env.assertEquals(res.properties_set, 6)
+        self.env.assertEquals(res.nodes_deleted, 0)
+
+        # validate that the correct nodes were created
+        n1 = Node(label = 'TEMP',
+                  properties={'v': 7, 'name': 'raz', 'li': [1, 2, 3, 4, 3]})
+        n2 = Node(label = 'TEMP',
+                  properties={'v': 9, 'name': 'raz', 'li': [1, 2, 3, 4, 4]})
+        res = self.get_res_and_assertEquals("""MATCH (t:TEMP) RETURN t ORDER BY
+                                            t.v""",
+                                            [[n1], [n2]])
+
+        # ----------------------------------------------------------------------
+        # triple embedded Foreach clause followed by reading and writing clauses
+        # ----------------------------------------------------------------------
+
+        # clear db
+        self.env.flush()
+        # Make a new graph object with no cache (problematic label info)
+        graph = Graph(self.env.getConnection(), GRAPH_ID)
+
+        # create a single node
+        graph.query("CREATE (:N {v: 1, name: 'RAZ', li: [1, 2, 3, 4]})")
+
+        query = """
+        MATCH (n:N)
+        WITH collect(n) as ns, toLower(n.name) as name
+        FOREACH(n in ns |
```

Comment:
make sure `FOREACH` array expression doesn't include an aggregation function e.g.
`FOREACH(a in collect(b) | ...)`